### ISSUE TITLE
BUG REPORT: docker logs follow gets stuck when the json-file log driver rotates files

### ISSUE THREAD

**Issue by lukeck** [docker/docker#23913]
**Output of `docker version`:**

```
Docker version 1.11.2, build b9f10c9
```

**Output of `docker info`:**

```
Containers: 5
 Running: 4
 Paused: 0
 Stopped: 1
Images: 8
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 63
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-24-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.795 GiB
Name: ip-10-47-10-249
ID: VVQ3:QEHD:WM5H:SY6Q:Y43P:MFQ6:BZ5Z:GO2M:YZ5Y:VQ5P:FSXW:JFDE
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Reproduced on Ubuntu 15.10, 16.04 EC2 instances. Reproduced on Ubuntu 14.04 docker-machine VM on my laptop. All tests were performed at a terminal in an SSH session connecting to a local docker daemon.

**Steps to reproduce the issue:**
1. To demonstrate the issue, Configure the Docker daemon to use the `json-file` log driver with a max-size of 10mb.
Mine looks like:
```bash
$ cat /etc/systemd/system/docker.service.d/docker-opts.conf
[Service]
ExecStart=
ExecStart=/usr/bin/docker daemon -H fd:// -s aufs --bip=192.168.255.1/24 --log-driver="json-file" --log-opt max-size=10m --log-opt max-file=5
```

1. Attempt to print the numbers 1 to 200000 to the stdout of a container and use `docker logs` to follow the output by running the following command:
```bash
$ docker run -d --name loggen ubuntu:16.04 /bin/bash -c 'for ((i=1; i<=200000; ++i)); do echo $i; done' && docker logs --timestamps --follow loggen > boom.txt
71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a
```

**Describe the results you received:**
I would expect to see the numbers 1 to 200000 along with the time each number was output by the `loggen` container in `boom.txt`

What I actually see:
```bash
$ tail boom.txt
2016-06-17T06:39:55.123832147Z 131491
2016-06-17T06:39:55.123837123Z 131492
2016-06-17T06:39:55.123842067Z 131493
2016-06-17T06:39:55.123846929Z 131494
2016-06-17T06:39:55.123851785Z 131495
2016-06-17T06:39:55.123856759Z 131496
2016-06-17T06:39:55.123861681Z 131497
2016-06-17T06:39:55.123866606Z 131498
2016-06-17T06:39:55.123871552Z 131499
2016-06-17T06:39:55.123876490Z 131500
```

Looking at the files generated by the json-file log driver for this container shows that things have gone wrong when the docker log driver rotates the json files.

```
$ sudo -s
$ ls /var/lib/docker/containers/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a/*.log*
/var/lib/docker/containers/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a-json.log
/var/lib/docker/containers/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a-json.log.1
$ tail /var/lib/docker/containers/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a-json.log.1
{"log":"131491\n","stream":"stdout","time":"2016-06-17T06:39:55.123832147Z"}
{"log":"131492\n","stream":"stdout","time":"2016-06-17T06:39:55.123837123Z"}
{"log":"131493\n","stream":"stdout","time":"2016-06-17T06:39:55.123842067Z"}
{"log":"131494\n","stream":"stdout","time":"2016-06-17T06:39:55.123846929Z"}
{"log":"131495\n","stream":"stdout","time":"2016-06-17T06:39:55.123851785Z"}
{"log":"131496\n","stream":"stdout","time":"2016-06-17T06:39:55.123856759Z"}
{"log":"131497\n","stream":"stdout","time":"2016-06-17T06:39:55.123861681Z"}
{"log":"131498\n","stream":"stdout","time":"2016-06-17T06:39:55.123866606Z"}
{"log":"131499\n","stream":"stdout","time":"2016-06-17T06:39:55.123871552Z"}
{"log":"131500\n","stream":"stdout","time":"2016-06-17T06:39:55.12387649Z"}
$ head -n 10 /var/lib/docker/containers/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a/71943ba916fb365fb3178ab9dabd37158d690b49aeaa595420fde4f92b62256a-json.log
{"log":"131501\n","stream":"stdout","time":"2016-06-17T06:39:55.123881404Z"}
{"log":"131502\n","stream":"stdout","time":"2016-06-17T06:39:55.123994259Z"}
{"log":"131503\n","stream":"stdout","time":"2016-06-17T06:39:55.124001593Z"}
{"log":"131504\n","stream":"stdout","time":"2016-06-17T06:39:55.124006555Z"}
{"log":"131505\n","stream":"stdout","time":"2016-06-17T06:39:55.124011382Z"}
{"log":"131506\n","stream":"stdout","time":"2016-06-17T06:39:55.124016163Z"}
{"log":"131507\n","stream":"stdout","time":"2016-06-17T06:39:55.124020854Z"}
{"log":"131508\n","stream":"stdout","time":"2016-06-17T06:39:55.12402562Z"}
{"log":"131509\n","stream":"stdout","time":"2016-06-17T06:39:55.124030396Z"}
{"log":"131510\n","stream":"stdout","time":"2016-06-17T06:39:55.12403511Z"}
```

**Describe the results you expected:**
I would expect to see the numbers 1 to 200000 along with the time each number was output by the `loggen` container in `boom.txt`

**Additional information you deem important (e.g. issue happens only occasionally):**
I did a little more digging and found an upper bound on the `--log-opt max-size` where the `docker logs --follow` was not broken by rotating the json log files.
When max-size is <= 215k, `docker logs --follow` behaves as expected.
When max-size is >= 216k, `docker logs --follow` gets stuck.

**Comment by cpuguy83** [comment]
Are you saying that the docker log follower didn't jump to the next log file?

**Comment by lukeck** [comment]
Yes. It stops following when logs start being sent to a second file.

**Comment by markine** [comment]
I'm seeing the same behavior with:

*docker version*:
```
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64
```

*docker info*:
```
Containers: 17
Images: 249
Server Version: 1.9.1
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 283
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.16.0-4-amd64
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 2
Total Memory: 7.815 GiB
Name: XXXXXXXXXXXXXXXXX
ID: 5MWW:M4LM:3UIM:5UOC:7AQD:DSWA:MIAA:WBFF:EMHR:2DH6:ITQC:QBVW
WARNING: No swap limit support
```

**Comment by thaJeztah** [comment]
@cpuguy83 could this be a duplicate of https://github.com/docker/docker/issues/23913? It's also using rotation

**Comment by cpuguy83** [comment]
@markine There's been a number of fixes since 1.9.1, please test on 1.11.2 and let us know if you still have a problem.

**Comment by cpuguy83** [comment]
@thaJeztah Not sure, that's this issue number :)

**Comment by cpuguy83** [comment]
@lukeck Can you provide daemon logs when this occurs?

**Comment by thaJeztah** [comment]
@cpuguy83 hmf, wrong copy/paste again; meant this one: https://github.com/docker/docker/issues/21220

**Comment by paralin** [comment]
Funny you guys would be looking at this exactly when I run into the problem :)

**Comment by cpuguy83** [comment]
Could you please test with #24514? Thanks!

**Comment by markine** [comment]
@cpuguy83 Thank you, unfortunately I don't think I have the luxury of bumping the docker version arbitrarily in my Kubernetes environment. @lukeck would love to hear whether #24514 resolves the issue for you.

**Comment by lukeck** [comment]
Nice one @cpuguy83! I should be able to take #24514 for a spin in the next few hours.

**Comment by paralin** [comment]
@cpuguy83 Confirming, fixes issues on ARM64.

**Comment by paralin** [comment]
@cpuguy83 Not sure if it is related to upgrading to rc4 or otherwise, but after applying your patch I am getting a bug.

```
docker exec -it mycontainer sh
$ exit
.... hang
```

Any ideas?

**Comment by paralin** [comment]
@cpuguy83 Never mind, the error I reported here before is unrelated, I think +1 for merging this

**Comment by ahmetalpbalkan** [comment]
We hit this a well. It looks like this is available on v1.13.x tags.
